src/Tasks/Dependencies.cs

```diff
@@ -1,32 +1,65 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.Build.BackEnd;
 using System;
-using System.Collections;
+using System.Collections.Generic;
 
 namespace Microsoft.Build.Tasks
 {
     /// <summary>
     /// Represents a cache of inputs to a compilation-style task.
     /// </summary>
-    /// <remarks>On-disk serialization format, don't change field names or types or use readonly.</remarks>
-    [Serializable]
     internal class Dependencies
     {
         /// <summary>
-        /// Hashtable of other dependency files.
+        /// Dictionary of other dependency files.
         /// Key is filename and value is DependencyFile.
         /// </summary>
-        private Hashtable dependencies = new Hashtable();
+        internal Dictionary<string, DependencyFile> dependencies = new();
+
+        internal Dependencies() { }
+
+        internal Dependencies(ITranslator translator, Type t)
+        {
+            Translate(translator, t);
+        }
+
+        public void Translate(ITranslator translator, Type t)
+        {
+            translator.TranslateDictionary(ref dependencies, (ITranslator translator, ref DependencyFile dependency) =>
+            {
+                if (t == typeof(ResGenDependencies.ResXFile))
```

Comment:
Why pass and check the type explicitly? Could you instead do a pattern match on the runtime type?

Reply:
Great question—that's actually how I first implemented it, and it took me a bit to figure out why it was failing. For serializing to disk, yes, that works. For deserializing, dependency comes in as null, so it can be cast to either a ResXFile or a PortableLibraryFile. Only one of those is valid, however, so I need to pick the right one. Explicitly passing the type (or some equivalent information) is the only way I can get the requisite information for deserializing the right type.

Reply:
Can we make Dependencies implement ITranslatable and override it in derived classes?

Reply:
If it was the first instinct for both of us it deserves a comment please

Reply:
@rokonec The type is just Dictionary<string, DependencyFile> and it has to work for either Dictionary<string, ResXFile> or Dictionary<string, PortableLibraryFile>. The Dependencies class doesn't know which will be required, so the Translator can't either. To be honest, this whole class structure could have been dramatically simplified by just having ResGenDependencies have a Dictionary<string, ResXFile> and a Dictionary<string, PortableLibraryFile>. That would make this a lot cleaner but be a little outside the scope of this PR. That said, if you want me to do that, I'm happy deleting a lot of code 😉